src/operator/tensor/index_add_forward.cc

```diff
@@ -0,0 +1,132 @@
+        val_id[_j] = (val_shape[_j] == 1) ? 0 : a_tail_id[_j];
+      }
+      val_id[seg + ind_ndim - 1] = (val_shape[seg + ind_ndim - 1] == 1) ? 0 : i;
+      index_t val_dest = mxnet_op::dot(val_id, val_stride);
+      #pragma omp critical
+      {
+        out[id + _i] += val[val_dest];
+      }
+    }
+  }
+};
+
+template<typename xpu, typename DType>
+void IndexAddForwardCalc(mshadow::Stream<xpu> *s,
+                         const int ind_num, DType* out,
+                        const DType* val,
+                        const mshadow::Shape<MXNET_SPECIAL_MAX_NDIM> a_tail_shape,
+                        const mshadow::Shape<MXNET_SPECIAL_MAX_NDIM> a_pre_stride,
+                        const mshadow::Shape<MXNET_SPECIAL_MAX_NDIM> val_stride,
+                        const mshadow::Shape<MXNET_SPECIAL_MAX_NDIM> val_shape,
+                        const mshadow::Shape<MXNET_SPECIAL_MAX_NDIM> a_shape,
+                        const int a_tail_size,
+                        const int ind_ndim, const int* ind,
+                        const int a_ndim) {
+  using namespace mxnet_op;
+  using namespace mshadow;
+  int seg = MXNET_SPECIAL_MAX_NDIM - a_ndim;
+  Kernel<IndexAddForwardCPUKernel<DType>, xpu>::Launch(
+                                             s, ind_num, out, val,
+                                             a_tail_shape, a_pre_stride,
+                                             val_stride, val_shape, a_shape,
+                                             a_tail_size, ind_num,
+                                             ind_ndim, ind, a_ndim, seg);
+}
+
+
+
+NNVM_REGISTER_OP(_npx_index_add)
+.describe(R"code(This operators implements the "+=" mimic function.
+)code" ADD_FILELINE)
+.set_num_inputs(3)
+.set_num_outputs(1)
+.set_attr<nnvm::FListInputNames>("FListInputNames",
+  [](const NodeAttrs& attrs) {
+    return std::vector<std::string>{"a", "ind", "val"};
+  })
+.set_attr<mxnet::FInferShape>("FInferShape", IndexModifyOpShape)
+.set_attr<nnvm::FInferType>("FInferType", IndexModifyOpType)
+.set_attr<FCompute>("FCompute<cpu>", IndexAddOpForward<cpu>)
+.set_attr<FResourceRequest>("FResourceRequest",
+  [](const NodeAttrs& attrs) {
+    return std::vector<ResourceRequest>{ResourceRequest::kTempSpace};
+  })
+.set_attr<nnvm::FGradient>("FGradient",
+  [](const nnvm::ObjectPtr& n, const std::vector<nnvm::NodeEntry>& ograds) {
+      auto a_grad = MakeNode("_copy", n->attrs.name + "_backward_a",
+                              {ograds[0]}, nullptr, &n);
+      auto idx_grad = MakeNode("zeros_like", n->attrs.name + "_backward_indices",
+                              {n->inputs[1]}, nullptr, &n);
+      auto val_grad = MakeNode("_backward_index_add_val", n->attrs.name + "_backward_val",
+                              {ograds[0], n->inputs[1]}, nullptr, &n);
+      std::vector<nnvm::NodeEntry> ret;
+      ret.emplace_back(a_grad);
+      ret.emplace_back(idx_grad);
+      ret.emplace_back(val_grad);
+      return ret;
```

Comment:
@yzhliu I've suggested @JiangZhaoh to write the gradient op by reusing the forward op. This has multiple benefits: 
1) We can do more optimization in the graph pass
2) It's easy to support higher-order gradient.

I think we may add the document about such usage since not many people are aware of that.